File: frontend/src/lib/api/predict.js
import { request } from './client';

/**
 * FaceSense Vision Prediction API
 * Communicates with FastAPI prediction and model endpoints:
 * - POST /api/v1/predict/image
 * - GET  /api/v1/model/info
 */
export const predictApi = {
  /**
   * Performs face detection and emotion classification on an uploaded image.
   * @param {File|Blob} file Image file to analyze
   * @param {Object} options Optional query options
   * @param {number} [options.confidenceThreshold] Minimum confidence threshold (0.0 to 1.0)
   * @param {boolean} [options.allowFallback] If true, treats entire image as face when face detector detects no faces
   * @returns {Promise<{faces_detected: number, predictions: Array<{box: {x: number, y: number, w: number, h: number}, emotion: string, confidence: number, all_probabilities: Object}>, processing_time_ms: number}>}
   */
  predictImage: async (file, options = {}) => {
    const formData = new FormData();
    formData.append('file', file);

    const queryParams = new URLSearchParams();
    if (typeof options.confidenceThreshold === 'number') {
      queryParams.append('confidence_threshold', options.confidenceThreshold.toString());
    }
    if (typeof options.allowFallback === 'boolean') {
      queryParams.append('allow_fallback', options.allowFallback.toString());
    }

    const queryString = queryParams.toString() ? `?${queryParams.toString()}` : '';

    return request(`/predict/image${queryString}`, {
      method: 'POST',
      body: formData,
    });
  },

  /**
   * Retrieves current active model metadata, architecture details, and supported classes.
   * @returns {Promise<{model_name: string, architecture: string, num_classes: number, classes: string[], device: string, input_shape: number[], status: string, metrics: Object|null}>}
   */
  getModelInfo: async () => {
    return request('/model/info', {
      method: 'GET',
    });
  },
};

File: backend/app/api/v1/endpoints/predict.py
from typing import Optional
from fastapi import APIRouter, File, UploadFile, Query, Depends

from backend.app.schemas.predict import ImagePredictionResponse
from backend.app.services.prediction_service import prediction_service
from backend.app.core.upload_security import validate_image_file
from backend.app.core.rate_limit import rate_limit_predict

router = APIRouter()


@router.post("/image", response_model=ImagePredictionResponse, tags=["Prediction"], dependencies=[Depends(rate_limit_predict)])
async def predict_image(
    file: UploadFile = File(..., description="Image file (JPEG, PNG, etc.) to analyze"),
    confidence_threshold: Optional[float] = Query(
        None, ge=0.0, le=1.0, description="Minimum confidence threshold (0.0 to 1.0)"
    ),
    allow_fallback: bool = Query(
        False, description="If True, treats entire image as face when face detector detects no faces"
    ),
) -> ImagePredictionResponse:
    """
    Detects faces in the uploaded image and predicts facial expressions/emotions
    for each detected face.
    """
    contents = await validate_image_file(file)
    return prediction_service.predict_image(
        image_bytes=contents,
        confidence_threshold=confidence_threshold,
        allow_fallback=allow_fallback,
    )
